### PR TITLE
[CUDAGraph] GPT3-175B Pipeline Parallel Training with CUDAGraph using PipelineParallelMicroStepCallback 

### PR DESCRIPTION
**Pull request by eee4017**
### PR Category
Distributed Strategy

### PR Types
New features

### Description

This PR introduces significant enhancements and fixes to improve the functionality and debugging capabilities of pipeline parallel training in PaddlePaddle. The primary addition is the `PipelineParallelMicroStepCallback`, which allows for better management of hooks within pipeline parallel processes. This update is crucial for supporting CUDA Graph pipeline parallel training and includes several other improvements.

**Key Features**
1. **PipelineParallelMicroStepCallback**:
    - This new feature facilitates enhanced management of hooks within pipeline parallel processes.
    - It allows for registering callbacks at specific pipeline stages: `forward_begin`, `forward_end`, `backward_begin`, and `backward_end`.
    - This functionality is particularly important for `PipelineParallel`, where layers are divided into multiple chunks.
    - The addition supports various tasks, such as logging, monitoring, and dynamic parameter adjustments during pipeline execution.
    - This feature aligns with PaddlePaddle's design principles and addresses the specific needs of pipeline parallelism.

2. **Support for CUDA Graph Pipeline Parallel Training**:
    - The update is essential for enabling efficient pipeline parallel training with CUDA Graph.
    - It allows training of large models, such as GPT-3 175B, on 64 H100 GPUs using hybrid parallelism (Pipeline Parallel + Tensor Parallel + Sequence Parallel), achieving a 1.18x speedup in training performance.

2. **Worker Log Adjustment**:
    - Updated the worker log to be ranked instead of being node-specific. This change ensures that the worker logs of each node do not collapse into a single log, facilitating better debugging and clarity.

3. **Debug Tools and Fixes**:
    - Added several debug tools and fixed issues in the CUDA graphed layer, enhancing the overall debugging experience and reliability of CUDA Graph pipeline training.

**Minor Fixes**
- Adjusted the worker log to rank-based logging.
- Improved debug tools and fixed issues in the CUDA graphed layer.

For more information, please check https://github.com/NVIDIA/TransformerEngine/pull/957
Check #65092

### CODE REVIEW

**Comment by jeng1220** [comment]
You must have one RD (phlrain or luotao1 or Aurelius84) approval

File: python/paddle/distributed/launch/controllers/ps.py [hunk 1]
```diff
@@ -111,7 +111,7 @@ def _build_pod_with_args(self):
                 "POD_IP": self.ctx.node.ip,
             }
             e.update(_gloo_envs)
-            log_file = f"workerlog.{i}"
+            log_file = f"workerlog.{i + trainer_rank_offset}"
```
Comment (JZ-LIANG) [inline python/paddle/distributed/launch/controllers/ps.py:114]: 这个log 命名格式必须改嘛？
集群侧部分log 监控分析程序依赖 log 的结尾是 [0 ~ 7]， 如果这块需要更新需要两边同时对齐

Comment (eee4017) [reply python/paddle/distributed/launch/controllers/ps.py:114]: 多节点的时候，这个log会叠再一起，导致除错看不懂，所以想说改一下这个编号。单节点的行为应该是与以前一样的，不影响

Comment (eee4017) [reply python/paddle/distributed/launch/controllers/ps.py:114]: 这个CI-coverage的覆蓋不夠多，看起来蛮多是在这workerlog的部分，应该原本就没测

Comment (jeng1220) [reply python/paddle/distributed/launch/controllers/ps.py:114]: @tianshuo78520a ,
上述CI-coverage不夠的問題也是找你處理嗎?

Comment (tianshuo78520a) [reply python/paddle/distributed/launch/controllers/ps.py:114]: > @tianshuo78520a , 上述CI-coverage不夠的問題也是找你處理嗎?

已经处理

Comment (ForFishes) [reply python/paddle/distributed/launch/controllers/ps.py:114]: > 多节点的时候，这个log会叠再一起，导致除错看不懂，所以想说改一下这个编号。单节点的行为应该是与以前一样的，不影响

这里的多节点怎么理解？这个改动会改变日志保存行为吗？

Comment (eee4017) [reply python/paddle/distributed/launch/controllers/ps.py:114]: 多节点是指多台机器的时候。多台机器多卡的时候每台机器的第0个device都会有一样的log编号，导致多个机器的log叠再一起

**Comment by jeng1220** [comment]
@sneaxiy , @JZ-LIANG , @ForFishes , @tianshuo78520a 
所有測試都通過了，請問可以合併這PR了嗎?

**Review by ForFishes — APPROVED** [review]
LGTM

**Pull request merged**